src/cpp/core/include/core/algorithm/Set.hpp

```diff
@@ -0,0 +1,51 @@
+/*
+ * Set.hpp
+ *
+ * Copyright (C) 2009-2015 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+#ifndef CORE_ALGORITHM_SET_HPP
+#define CORE_ALGORITHM_SET_HPP
+
+// Convenience algorithms -- not necessarily as fast as just
+// using the STL + iterators, but often easier to use.
+
+#include <algorithm>
+
+namespace rstudio {
+namespace core {
+namespace algorithm {
+
+#define RS_SET_OPERATION(__OPERATION__)                                        \
```

Comment:
I honestly don't love the pattern of defining macros for commonly used code snippets (including the IF_ERROR one I saw you defined before). I think it makes the code less readable and less debuggable. I'd rather use functions even if they are a bit more verbose. As with all things though open to it though if there's a sound rationale. @jmcphers What's your take?